packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
+
+        // A stack of routes that are referenced through handles in DDSes.
+        const referencedRoutesStack: string[] = [];
+
+        // Start with the root data stores. Add all the fluid objects referenced by the root data stores in to the
+        // referencedObjectsStack.
+        this.contexts.forEach((value: FluidDataStoreContext) => {
+            if (value.isRootStore) {
+                // Ignore the scheduler.
+                if (value.id === taskSchedulerId) {
+                    return;
+                }
+
+                // Get the fluid objects referenced by this data store and add them to the referencedObjectsStack.
+                const referencedRoutes = value.getReferencedRoutes();
+                referencedRoutesStack.push(...referencedRoutes);
+
+                // Set the IReferences object in the dataStoreReferencesMap for this data store. Mark it as visited.
+                dataStoreReferencesMap.set(value.id, { visited: true, referencedRoutes: new Set<string>() });
+            }
+        });
+
+        // Do a Bread First Search on the referencedRoutesStack.
+        while (referencedRoutesStack.length > 0) {
+            // Get the next route.
+            const nextRoute = referencedRoutesStack.shift();
+            assert(nextRoute);
+
+            // Get the dataStoreId from the referenced object.
+            const dataStoreId = nextRoute.split("/")[1];
+            // If there is no reference object for it in the dataStoreReferencesMap, it means this is the first time we
+            // are seeing it. Creata a reference object for it.
+            let references = dataStoreReferencesMap.get(dataStoreId);
+            if (references === undefined) {
+                references = {
+                    visited: false,
+                    referencedRoutes: new Set<string>(),
+                };
+            }
+
+            const visited: boolean = references.visited;
+            // Keeps track of whether the entry for this data store in the dataStoreReferencesMap changes.
+            let dataStoreEntryChanged: boolean = false;
+
+            // Get the path of the referenced object relative to the dataStoreId and add it to the
+            // referencedFluidObjects in its references.
+            const dataStoreIdRelativePath = nextRoute.slice(dataStoreId.length + 1);
+            if (dataStoreIdRelativePath !== "") {
+                assert(references.referencedRoutes);
+                references.referencedRoutes.add(dataStoreIdRelativePath);
+                dataStoreEntryChanged = true;
+            }
+
+            // If this data store has not been visited yet, mark it as visited.
+            if (!references.visited) {
+                references.visited = true;
+                dataStoreEntryChanged = true;
+            }
+
+            // If the entry has changed, update it in the dataStoreReferencesMap.
+            if (dataStoreEntryChanged) {
```

Comment:
Can this update be unconditional (and drop dataStoreEntryChanged variable)?

Reply:
I believe so. I added this flag as I didn't want to unnecessary update the map. I can remove this.

Reply:
If update is a single property, then it's cheaper to do it then to do many writes to this variable and have IF statement. branching is the most expensive stuff for processor :)

Reply:
True, branching is expensive :-).